### PR TITLE
Support multiple instances in the same network

### PR DESCRIPTION
**Pull request by lentzi90**
- Add variables that can be used to configure dnsmasq to ignore certain hosts, e.g. unknown.
- Use PROVISIONING_IP variable instead of hard coded IP
- Use EXTERNAL_SUBNET_VX to set CLUSTER_APIENDPOINT_IP

With this in place it is possible to tell dnsmasq about the known hosts and then ignore all other hosts. For example:
DHCP_HOSTS=00:20:e0:3b:13:af;00:20:e0:3b:14:af
DHCP_IGNORE=tag:!known

Related ironic-image PR that adds the DHCP variables: https://github.com/metal3-io/ironic-image/pull/394
/hold

### CODE REVIEW

File: 03_launch_mgmt_cluster.sh [hunk 1]
```diff
@@ -146,6 +146,14 @@ EOF
     echo "IRONIC_KERNEL_PARAMS=console=ttyS0" | sudo tee -a "$IRONIC_DATA_DIR/ironic_bmo_configmap.env"
   fi
 
+  if [ -n "${DHCP_IGNORE}" ]; then
+    echo "DHCP_IGNORE=${DHCP_IGNORE}" | sudo tee -a "$IRONIC_DATA_DIR/ironic_bmo_configmap.env"
```
Comment (tuminoid) [inline 03_launch_mgmt_cluster.sh:150]: Missing curlys for IRONIC_DATA_DIR here and on line 154.

File: lib/network.sh [hunk 2]
```diff
@@ -118,6 +110,23 @@ else
...
+  # Accept user provided value if set
+  export CLUSTER_APIENDPOINT_IP
+elif [[ -n "${EXTERNAL_SUBNET_V4}" ]]; then
+  network_address CLUSTER_APIENDPOINT_IP "${EXTERNAL_SUBNET_V4}" 249
+else
+  network_address CLUSTER_APIENDPOINT_IP "${EXTERNAL_SUBNET_V6}" 249
+fi
+
+# shellcheck disable=SC2153
+if [[ "$CLUSTER_APIENDPOINT_IP" == *":"* ]]; then
```
Comment (tuminoid) [inline lib/network.sh:123]: This one is missing curlys around the variable, vs the others above it. Lines 124 and 126 have the same as well.

**Review by tuminoid — COMMENTED** [review]
Nitpicking on the inconsistent use of curlys.

File: lib/network.sh [hunk 3]
```diff
@@ -118,6 +110,23 @@ else
     exit 1
 fi
 
+if [[ -n "${CLUSTER_APIENDPOINT_IP}" ]]; then
+  # Accept user provided value if set
+  export CLUSTER_APIENDPOINT_IP
+elif [[ -n "${EXTERNAL_SUBNET_V4}" ]]; then
+  network_address CLUSTER_APIENDPOINT_IP "${EXTERNAL_SUBNET_V4}" 249
+else
+  network_address CLUSTER_APIENDPOINT_IP "${EXTERNAL_SUBNET_V6}" 249
```
Comment (furkatgofurov7) [inline lib/network.sh:119]: I was not aware of this ipv6 IP being used for v6 family.., though we never test this in the CI and it looks to be not allocated in other places: https://github.com/metal3-io/metal3-dev-env/blob/681b4d26ab5f90b0d08679df36b82ebb74eca993/lib/network.sh#L152 , https://github.com/metal3-io/metal3-dev-env/blob/681b4d26ab5f90b0d08679df36b82ebb74eca993/lib/network.sh#L122

File: vars.md [hunk 4]
```diff
@@ -111,6 +111,8 @@ assured that they are persisted.
 | BUILD_CAPI_LOCALLY | build Cluster API based on CAPIPATH | "true","false" | "false" |
 | BUILD_IRONIC_IMAGE_LOCALLY | build the Metal3's ironic-image based on IRONIC_IMAGE_PATH | "true","false" | "false" |
 | IRONIC_FROM_SOURCE | installs ironic from source during container image building, if `true` then the `BUILD_IRONIC_IMAGE_LOCALLY` will be also set to `true` | "true","false" | "false" |
+| DHCP_HOSTS | A list of `;` separated dhcp-host directives for dnsmasq | e.g. `00:20:e0:3b:13:af;00:20:e0:3b:14:af` | |
+| DHCP_IGNORE | A set of tags on hosts to be ignored by dnsmasq | e.g. `tag:!known` | |
```
Comment (furkatgofurov7) [inline vars.md:115]: are these related to this series of kustomization PRs?

Comment (lentzi90) [reply vars.md:115]: No this is for running multiple instances in the same network. Useful for example in a baremetal lab where multiple hosts are on the same network. Then we can simply list the MACs that we care about and tell dnsmasq to ignore the rest.

Comment (lentzi90) [reply vars.md:115]: It is related this PR that I also linked in the description: https://github.com/metal3-io/ironic-image/pull/394

Comment (furkatgofurov7) [reply vars.md:115]: ah sorry, I confused this PR with https://github.com/metal3-io/metal3-dev-env/pull/1086

**Comment by tuminoid** [comment]
No more nits from me.

File: lib/network.sh (hunk 3 above)
Comment (furkatgofurov7) [reply lib/network.sh:119]: It was more of a like question / observation point I had, and not a blocker. I was also assuring that 249 IP is not used anywhere so it is safe to assign it.

File: 03_launch_mgmt_cluster.sh [hunk 5]
```diff
@@ -139,15 +139,23 @@ IRONIC_RAMDISK_SSH_KEY=${SSH_PUB_KEY_CONTENT}
...
-    echo "IRONIC_KERNEL_PARAMS=console=ttyS0" | sudo tee -a "$IRONIC_DATA_DIR/ironic_bmo_configmap.env"
+    echo "IRONIC_KERNEL_PARAMS=console=ttyS0" | sudo tee -a "${IRONIC_DATA_DIR}/ironic_bmo_configmap.env"
+  fi
+
+  if [ -n "${DHCP_IGNORE}" ]; then
+    echo "DHCP_IGNORE=${DHCP_IGNORE}" | sudo tee -a "${IRONIC_DATA_DIR}/ironic_bmo_configmap.env"
+  fi
+
+  if [ -n "${DHCP_HOSTS}" ]; then
+    echo "DHCP_HOSTS=${DHCP_HOSTS}" | sudo tee -a "$IRONIC_DATA_DIR/ironic_bmo_configmap.env"
```
Comment (smoshiur1237) [inline 03_launch_mgmt_cluster.sh:154]: Curlys are missing in IRONIC_DATA_DIR

Comment (lentzi90) [reply 03_launch_mgmt_cluster.sh]: Thanks, fixed!

**Review by smoshiur1237 — APPROVED** [review]
/lgtm

File: tests/roles/run_tests/vars/main.yml [hunk 6]
```diff
@@ -39,6 +39,7 @@ BAREMETALV6_POOL_RANGE_START: "{{ lookup('env', 'BAREMETALV6_POOL_RANGE_START')
 BAREMETALV6_POOL_RANGE_END: "{{ lookup('env', 'BAREMETALV6_POOL_RANGE_END') | default('fd55::200', true) }}"
 EXTERNAL_SUBNET_V6_PREFIX: "{{ lookup('env', 'EXTERNAL_SUBNET_V6_PREFIX') | default('64', true) }}"
 EXTERNAL_SUBNET_V6_HOST: "{{ lookup('env', 'EXTERNAL_SUBNET_V6_HOST') | default('fd55::1', true) }}"
+PROVISIONING_IP: "{{ lookup('env', 'PROVISIONING_IP') | default('172.22.0.1', true) }}"
```
Comment (kashifest) [inline tests/roles/run_tests/vars/main.yml:42]: Do we need to add it in lib/common.sh as other vars or no need?

Comment (lentzi90) [reply tests/roles/run_tests/vars/main.yml:42]: I would say no. It is in lib/network.sh and that should be enough. I just added it here in the ansible vars now so it will also respect it if we set a different value. Previously we had a hard coded value in ansible.

Comment (kashifest) [reply tests/roles/run_tests/vars/main.yml:42]: ah missed that, cool, thats enough

**Review by kashifest — APPROVED** [review]
